core/lib/state/src/shadow_storage.rs

```diff
@@ -24,55 +24,73 @@ static METRICS: vise::Global<ShadowStorageMetrics> = vise::Global::new();
 /// to_check_storage -- secondary storage, which will verify it's own return values against source_storage
 /// Note that if to_check_storage value is different than source value, execution continues and metrics/ logs are emitted.
 #[derive(Debug)]
-pub struct ShadowStorage<'a> {
-    source_storage: Box<dyn ReadStorage + 'a>,
-    to_check_storage: Box<dyn ReadStorage + 'a>,
-    metrics: &'a ShadowStorageMetrics,
+pub struct ShadowStorage<Ref, Check> {
+    source_storage: Ref,
+    to_check_storage: Check,
+    metrics: &'static ShadowStorageMetrics,
     l1_batch_number: L1BatchNumber,
+    panic_on_divergence: bool,
 }
 
-impl<'a> ShadowStorage<'a> {
+impl<Ref, Check> ShadowStorage<Ref, Check> {
     /// Creates a new storage using the 2 underlying [`ReadStorage`]s, first as source, the second to be checked
     /// against the source.
     pub fn new(
-        source_storage: Box<dyn ReadStorage + 'a>,
-        to_check_storage: Box<dyn ReadStorage + 'a>,
+        source_storage: Ref,
+        to_check_storage: Check,
         l1_batch_number: L1BatchNumber,
     ) -> Self {
         Self {
             source_storage,
             to_check_storage,
             metrics: &METRICS,
             l1_batch_number,
+            panic_on_divergence: true,
         }
     }
 }
 
-impl ReadStorage for ShadowStorage<'_> {
+impl<Ref: ReadStorage, Check: ReadStorage> ReadStorage for ShadowStorage<Ref, Check> {
     fn read_value(&mut self, key: &StorageKey) -> StorageValue {
-        let source_value = self.source_storage.as_mut().read_value(key);
-        let expected_value = self.to_check_storage.as_mut().read_value(key);
+        let source_value = self.source_storage.read_value(key);
+        let expected_value = self.to_check_storage.read_value(key);
         if source_value != expected_value {
             self.metrics.read_value_mismatch.inc();
-            tracing::error!(
-                "read_value({key:?}) -- l1_batch_number={:?} -- expected source={source_value:?} \
-                 to be equal to to_check={expected_value:?}",
-                self.l1_batch_number
-            );
+            if self.panic_on_divergence {
+                panic!(
+                    "read_value({key:?}) -- l1_batch_number={:?} -- expected source={source_value:?} \
+                     to be equal to to_check={expected_value:?}",
+                    self.l1_batch_number
+                );
+            } else {
+                tracing::error!(
+                    "read_value({key:?}) -- l1_batch_number={:?} -- expected source={source_value:?} \
+                     to be equal to to_check={expected_value:?}",
+                    self.l1_batch_number
+                );
+            }
```

Comment:
Maybe it would make the code cleaner if this was a function? Something like `error_or_panic(&str)` so you don't have to repeat formatting twice everywhere